### PR TITLE
Make category pages indexable

### LINKED ISSUE #10069: Make category pages indexable
We currently [forbid robots from indexing /{lang}/{app}/search/](https://addons.mozilla.org/robots.txt). We want to keep that, but at the same time, we're currently re-use the search pages for category pages, and we'd like _those_ to be indexed.

Easiest way to start fixing this is probably going to resolve around exposing "true" category pages under a specific URL, leaving the functionality and underlying templates/API calls the same. Eventually we may want to completely customize the category pages to make them look different, add specific content etc but that will be different follow-up issue(s).

Note: we also forbid indexing collections, but explicitly allow the Mozilla ones by id and username, so all the "special" collections are currently indexed already.

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #10069

### CODE REVIEW

File: src/amo/components/Routes/index.js [hunk 1]
```diff
@@ -147,6 +147,12 @@ const Routes = ({ _config = config }: Props = {}) => (
       component={CategoriesPage}
     />
 
+    <Route
+      exact
+      path="/:lang/:application/category/"
```
Comment (willdurand) [inline src/amo/components/Routes/index.js:152]: We probably want per-category URLs so it should be something like this:

```
path="/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/"
```

File: src/amo/components/Routes/index.js [hunk 2]
```diff
@@ -147,6 +147,12 @@ const Routes = ({ _config = config }: Props = {}) => (
...
+      component={(props) => <SearchPage {...props} isForCategory />}
```
Comment (willdurand) [inline src/amo/components/Routes/index.js:153]: It'd be best to create a new page component (`CategoryPage`?). We'll want to use `HeadLinks` and maybe `HeadMetaTags` in the future.

File: src/amo/components/Routes/index.js (hunk 1 above)
Comment (bobsilverberg) [reply src/amo/components/Routes/index.js:152]: Why do we need those? I thought this was just about the ability to crawl. I can change it, but this is the most expedient way to accomplish this.

File: src/amo/components/Routes/index.js (hunk 2 above)
Comment (bobsilverberg) [reply src/amo/components/Routes/index.js:153]: Again, I'm not sure if we need these if the only point is to have a page for the crawler to crawl. Will the page be penalized if it doesn't contain those?

Comment (bobsilverberg) [reply src/amo/components/Routes/index.js:153]: FWIW, I did start with a new page component and then realized "hey, we don't need that", but I guess maybe we do.

File: src/amo/components/Routes/index.js (hunk 1 above)
Comment (willdurand) [reply src/amo/components/Routes/index.js:152]: Pretty urls are a bit better in terms of crawling experience and we're going to enhance these pages in the future. I think it's best to plan for the future, it's not a lot of work.

We can mimic what is done in `SearchPage`. We don't need to worry about the server redirects, etc. so the `CategoryPage` should be a simple component (famous last words?)

File: src/amo/components/Routes/index.js (hunk 2 above)
Comment (willdurand) [reply src/amo/components/Routes/index.js:153]: Yeah, I think we do need a new page component. We need `HeadLinks` to tell the crawlers what to crawl.

**Comment by bobsilverberg** [comment]
We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?

File: src/amo/components/Search/index.js [hunk 3]
```diff
@@ -115,100 +116,104 @@ export class SearchBase extends React.Component<InternalProps> {
   }
 
   renderHelmetTitle() {
-    const { i18n, filters } = this.props;
+    const { i18n, filters, pageTitle } = this.props;
 
-    let title = i18n.gettext('Search results');
+    let title = pageTitle;
 
-    if (filters.promoted) {
```
Comment (bobsilverberg) [inline src/amo/components/Search/index.js:122]: FYI, the way GitHub has interpreted this diff, makes it looks like there are a lot of changes, but in fact the code below is completely unchanged. The only change is to introduce the use of the `pageTitle` prop as an override.

**Comment by bobsilverberg** [comment]
@willdurand This is ready for another look.

**Comment by willdurand** [comment]
> We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?

do we?

**Comment by bobsilverberg** [comment]
> > We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?
> 
> do we?

I recall having to do it before, and also, there's this comment: https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42, but maybe I'm wrong?

@bqbn Do we need anything in nginx to support a new route in our app? We've added a route for `path="/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/"`.

**Comment by willdurand** [comment]
> I recall having to do it before, and also, there's this comment: [https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42](https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js?rgh-link-date=2021-03-04T19%3A34%3A21Z#L42), but maybe I'm wrong?

I was looking at this comment. Given that the sentence doesn't really seem written in English, I guess I wrote it? 😅

**Comment by bobsilverberg** [comment]
> > I recall having to do it before, and also, there's this comment: [https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42](https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js?rgh-link-date=2021-03-04T19%3A36%3A00Z#L42), but maybe I'm wrong?
> 
> I was looking at this comment. Given that the sentence doesn't really seem written in English, I guess I wrote it? 😅

Blame agrees 😉

**Comment by bqbn** [comment]
> > > We also need to have ops add a route to nginx for this new route. Is that generally done via a bug or via an issue?
> > 
> > 
> > do we?
> 
> I recall having to do it before, and also, there's this comment: https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Routes/index.js#L42, but maybe I'm wrong?
> 
> @bqbn Do we need anything in nginx to support a new route in our app? We've added a route for `path="/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/"`.

Please take a look at [this comment](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L352). Based on the 4 tickets mentioned in that comment, there are 4 existing redirection rules that deal with the `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/` path. Please make sure that doesn't conflict with what you're trying to achieve here.

That said, by default, Nginx routes everything under `/` to the frontend.

**Comment by willdurand** [comment]
> Please take a look at [this comment](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L352). Based on the 4 tickets mentioned in that comment, there are 4 existing redirection rules that deal with the `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/` path. Please make sure that doesn't conflict with what you're trying to achieve here.

ugh, that doesn't sound good.

**Comment by bobsilverberg** [comment]
@jvillalobos Because of some permanent redirects put in place back in May or 2019, we cannot reuse the old URL path for category pages, which was `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/`. We need some alternate variation. A few suggestions are:

- `/:lang/:application/:visibleAddonType(extensions|themes)/category/:categorySlug/`
- `/:lang/:application/:visibleAddonType(extensions|themes)/c/:categorySlug/`
- `/:lang/:application/c/:visibleAddonType(extensions|themes)/:categorySlug/`

What do you think/recommend?

**Comment by bqbn** [comment]
Why can't we remove [the 4 redirection rules](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L353-L366) from Nginx config?

**Comment by bobsilverberg** [comment]
> Why can't we remove [the 4 redirection rules](https://github.com/mozilla-services/cloudops-deployment/blob/dc76521ac40e8c5b29c0d50208c98a793c2dbda4/projects/amo/puppet/modules/amo_proxy/templates/nginx.addons.conf.erb#L353-L366) from Nginx config?

If we remove the redirects then when someone does visit one of those routes it will be a 404, because the new route for the category page will not follow the pattern of `/:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/`, unless I am mistaken.

**Comment by jvillalobos** [comment]
> /:lang/:application/:visibleAddonType(extensions|themes)/category/:categorySlug/

That one looks fine, but I'm also not clear about the redirects.

> the new route for the category page will not follow the pattern of /:lang/:application/:visibleAddonType(extensions|themes)/:categorySlug/

Aren't we essentially trying to bring back the old URLs? Why wouldn't it work to drop the redirects?

**Comment by willdurand** [comment]
> Aren't we essentially trying to bring back the old URLs? Why wouldn't it work to drop the redirects?

We can't, that would lead to redirect loops in some cases because existing redirects are permanent. This is why we need a new URL pattern.

File: src/amo/pages/CategoryPage/index.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  i18n: I18nType,
+|};
+
+export class CategoryPageBase extends React.Component<InternalProps> {
+  getPageTitle(addonType: string, categoryName: string | null) {
+    const { i18n } = this.props;
+
+    // The only two possibilites for addonType are 'extension' and
+    // 'statictheme' because anything else would have thrown an
+    // exception in mapStateToProps via apiAddonType()
```
Comment (willdurand) [inline src/amo/pages/CategoryPage/index.js:53]: The router should reject invalid values because the `path` is a regular expression.

File: src/amo/pages/CategoryPage/index.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    if (addonType === ADDON_TYPE_EXTENSION) {
```
Comment (willdurand) [inline src/amo/pages/CategoryPage/index.js:54]: I'd rather special case themes because we could _potentially_ see a theme as an extension but the opposite isn't true.

File: src/amo/pages/CategoryPage/index.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      this.props.dispatch(
+        fetchCategories({ errorHandlerId: this.props.errorHandler.id }),
+      );
+    }
+  }
+
+  render() {
+    const { categoryName, filters, match } = this.props;
+    const { categorySlug, visibleAddonType } = match.params;
+    const addonType = apiAddonType(visibleAddonType);
```
Comment (willdurand) [inline src/amo/pages/CategoryPage/index.js:82]: This could be injected in `mapStateToProps`.

File: src/amo/pages/CategoryPage/index.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            addonType,
+            slug: categorySlug,
+          })}
+        />
+      </Page>
+    );
+  }
+}
+
+function mapStateToProps(state: AppState, ownProps: Props): PropsFromState {
```
Comment (willdurand) [inline src/amo/pages/CategoryPage/index.js:107]: Although the codebase is still inconsistent, we tend to prefer arrow functions.

File: src/amo/pages/CategoryPage/index.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+  if (categoriesState && clientApp) {
+    const appTypes = categoriesState[clientApp];
+
+    if (appTypes) {
+      const categories = appTypes[addonType];
+
+      if (categories && categories[categorySlug]) {
+        categoryName = categories[categorySlug].name;
+      }
```
Comment (willdurand) [inline src/amo/pages/CategoryPage/index.js:125]: This logic could be moved into a helper function.

File: tests/unit/amo/pages/TestSearchPage.js [hunk 9]
```diff
@@ -153,6 +157,28 @@ describe(__filename, () => {
     sinon.assert.notCalled(fakeDispatch);
   });
 
+  it('dispatches a server redirect when `category` and `addonType` is set', () => {
```
Comment (willdurand) [inline tests/unit/amo/pages/TestSearchPage.js:160]: why is it "is set" and not "are set"?

File: tests/unit/amo/pages/TestSearchPage.js [hunk 10]
```diff
@@ -153,6 +157,28 @@ describe(__filename, () => {
...
+      }),
+      store,
+    });
+
+    sinon.assert.callCount(fakeDispatch, 1);
+    sinon.assert.calledWith(
+      fakeDispatch,
+      sendServerRedirect({
+        status: 301,
+        url: `/en-US/android/extensions/category/${category}/?page=${page}`,
```
Comment (willdurand) [inline tests/unit/amo/pages/TestSearchPage.js:177]: Maybe explicitly pass the `lang` or `clientApp` in this test case to make sure the values are updated correctly.

**Review by willdurand — COMMENTED** [review]
Some comments while I am testing the feature locally

File: src/amo/pages/SearchTools/index.js [hunk 11]
```diff
@@ -3,85 +3,50 @@ import * as React from 'react';
...
+      addonType: ADDON_TYPE_EXTENSION,
+      slug: 'search-tools',
+    });
 
     dispatch(
       sendServerRedirect({
         status: 301,
-        url: `/${lang}/${clientApp}/search/${queryString}`,
+        url: `/${lang}/${clientApp}${pathname}${makeQueryString({
+          sort: SEARCH_SORT_TOP_RATED,
```
Comment (willdurand) [inline src/amo/pages/SearchTools/index.js:35]: mm it isn't the same thing as before, right? I don't see a `sort` filter in the previous code?

File: src/amo/components/SearchFilters/index.js [hunk 12]
```diff
@@ -105,15 +105,24 @@ export class SearchFiltersBase extends React.Component<InternalProps> {
...
+    if (filters.page) {
       // Since it's now a new search, reset the page.
       // eslint-disable-next-line
-      newFilters.page = '1';
+      filters.page = '1';
+    }
+
+    // If category is a filter, remove category and type as they are already
+    // included in pathname.
+    if (filters.category) {
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/SearchFilters/index.js:118]: Good point. It shouldn't ever be needed.

File: src/amo/pages/CategoryPage/index.js (hunk 4 above)
Comment (bobsilverberg) [reply src/amo/pages/CategoryPage/index.js:53]: That's true, but the result is the same. This comment is explaining why we don't need to address any different add-on types. I guess I can just remove the comment, or would you prefer that I update it to explain your reason why we don't need to worry about other cases?

File: src/amo/pages/CategoryPage/index.js (hunk 6 above)
Comment (bobsilverberg) [reply src/amo/pages/CategoryPage/index.js:82]: I started by doing that and it was creating an issue/complication, but I don't remember what it was. I can try again.

File: src/amo/pages/CategoryPage/index.js (hunk 8 above)
Comment (bobsilverberg) [reply src/amo/pages/CategoryPage/index.js:125]: It looks like it's only used once, here, so what's the advantage? Just to make the `mapStateToProps` function smaller/cleaner?

Comment (bobsilverberg) [reply src/amo/pages/CategoryPage/index.js:125]: Oh, wait, it's also in `SearchContextCard`, just using different variable names. 😛

File: tests/unit/amo/pages/TestSearchPage.js (hunk 9 above)
Comment (bobsilverberg) [reply tests/unit/amo/pages/TestSearchPage.js:160]: Because my English is poor (or I wasn't paying attention).

**Comment by bobsilverberg** [comment]
Note that not all of the comments have been addressed yet, but I am eoding, so I'm going to push a commit that addresses most of them.

File: src/amo/pages/SearchTools/index.js (hunk 11 above)
Comment (bobsilverberg) [reply src/amo/pages/SearchTools/index.js:42]: It was in `mapStateToProps` in the old code, but looking at the current behaviour of the site it isn't being picked up, so I guess we can remove it.

**Comment by bobsilverberg** [comment]
I have now addressed all of the comments. I did find one regression when testing though, and I don't have time to address it this afternoon, but I should be able to deal with it next week.

If you are on the new CategoryPage and you type a query into the Search input in the header, it no longer searches within the category. I believe this is because the category is no longer in the URL as a query param, but I am guessing there is another way for us to address this, other than adding it back in. I'll look into it next week.

**Comment by willdurand** [comment]
> If you are on the new CategoryPage and you type a query into the Search input in the header, it no longer searches within the category. I believe this is because the category is no longer in the URL as a query param, but I am guessing there is another way for us to address this, other than adding it back in. I'll look into it next week.

Given that we are not on a "search results" page anymore, do we want to keep this behavior?

**Comment by bobsilverberg** [comment]
> > If you are on the new CategoryPage and you type a query into the Search input in the header, it no longer searches within the category. I believe this is because the category is no longer in the URL as a query param, but I am guessing there is another way for us to address this, other than adding it back in. I'll look into it next week.
> 
> Given that we are not on a "search results" page anymore, do we want to keep this behavior?

Hmm, maybe not? What do you think @jvillalobos ?

**Comment by jvillalobos** [comment]
Honestly I didn't remember we did this 😀. It's a nice feature, but not really discoverable. I think it should be fine to remove it.

**Comment by bobsilverberg** [comment]
@willdurand I'd really like to get this into this week's tag, so if you have time to take another look today, or tomorrow morning, that would be great.

**Comment by willdurand** [comment]
> @willdurand I'd really like to get this into this week's tag, so if you have time to take another look today, or tomorrow morning, that would be great.

So this is ready for another look? No more regression to fix?

**Comment by bobsilverberg** [comment]
> > @willdurand I'd really like to get this into this week's tag, so if you have time to take another look today, or tomorrow morning, that would be great.
> 
> So this is ready for another look? No more regression to fix?

Correct, that's why I asked. ;-)  Jorge seems fine with the one "regression" listed above. I didn't find any others.

File: src/amo/pages/CategoryPage/index.js [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    addonType,
+    categoryName,
+    filters: fixFiltersFromLocation(filtersFromLocation),
+  };
+};
+
+const CategoryPage: React.ComponentType<Props> = compose(
+  connect(mapStateToProps),
+  translate(),
+  withErrorHandler({ name: 'CategoryPage' }),
```
Comment (willdurand) [inline src/amo/pages/CategoryPage/index.js:143]: Shouldn't we be using a "fixed" error handler given that it's a page-level component?

File: src/amo/components/SearchFilters/index.js [hunk 14]
```diff
@@ -114,15 +114,17 @@ export class SearchFiltersBase extends React.Component<InternalProps> {
   doSearch(newFilters: SearchFiltersType) {
     const { clientApp, lang, history, pathname } = this.props;
 
-    if (newFilters.page) {
```
Comment (willdurand) [inline src/amo/components/SearchFilters/index.js:117]: Why is this change needed actually?

File: src/amo/pages/SearchPage/index.js [hunk 15]
```diff
@@ -43,9 +45,23 @@ export class SearchPageBase extends React.Component<InternalProps> {
...
 
+    // If this is an old category search, redirect to the new category page.
+    if (newFilters.category && newFilters.addonType) {
+      pathname = getCategoryResultsPathname({
+        addonType: newFilters.addonType,
+        slug: newFilters.category,
+      });
+
+      delete newFilters.addonType;
+      delete newFilters.category;
```
Comment (willdurand) [inline src/amo/pages/SearchPage/index.js:60]: I think we should also delete `sort` if it's equal to the default sort because we should redirect to the canonical category url when possible.

File: src/amo/pages/CategoryPage/index.js [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+)(CategoryPageBase);
+
+export default CategoryPage;
```
Comment (willdurand) [inline src/amo/pages/CategoryPage/index.js:146]: Shouldn't we return a 404 when the category slug is invalid? For example: http://localhost:3000/en-US/firefox/themes/category/nope/

**Review by willdurand — COMMENTED** [review]
I have a few more comments.

File: src/amo/components/SearchFilters/index.js (hunk 14 above)
Comment (bobsilverberg) [reply src/amo/components/SearchFilters/index.js:117]: It's not, anymore. Subsequent changes have rendered it unnecessary. Thanks for pointing it out.

File: src/amo/pages/CategoryPage/index.js (hunk 16 above)
Comment (bobsilverberg) [reply src/amo/pages/CategoryPage/index.js:146]: Good catch. I assumed the logic in `Page` would take care of that, but the API error is coming from the `Search` component. I think the best way to address this is to just change the logic of the `Search` component to display a 404 page if the API response is either `404` or `400`. I can't think of an occasion on which we'd want to display anything other than a 404 page when the search API returns a 400. I've made this change and you can see what you think when I push the commit.

**Comment by bobsilverberg** [comment]
Hmm, I'm seeing a new/old issue. When you change the sort order from the SearchFilters component on the CategoryPage, it adds the `category` and `type` to the URL. I know I had code to remove those before, but now they're back. I think maybe there was a comment directing me to remove the code I had for that, but now I think we need it again.

File: src/amo/components/SearchFilters/index.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/SearchFilters/index.js]: I think this might be the comment that made me remove the code that removes the category and type filters, but I think we did in fact need that code, as it was used to remove them when a user changes the sort sequence from the SearchFilters component. I'll add it back in and see if that fixes it.

File: src/amo/components/SearchFilters/index.js (hunk 14 above)
Comment (bobsilverberg) [reply src/amo/components/SearchFilters/index.js:117]: And... this code is back now 😉

**Comment by bobsilverberg** [comment]
I fixed the issue with the category and type being added to the URL by bringing back the code I wrote earlier for `SearchFilters`.

**Comment by bobsilverberg** [comment]
@willdurand This is ready for another look.

File: src/amo/components/Search/index.js [hunk 17]
```diff
@@ -251,7 +256,7 @@ export class SearchBase extends React.Component<InternalProps> {
     if (errorHandler.hasError()) {
       log.warn(`Captured API Error: ${errorHandler.capturedError.messages}`);
 
-      if (errorHandler.capturedError.responseStatusCode === 404) {
+      if ([400, 404].includes(errorHandler.capturedError.responseStatusCode)) {
```
Comment (willdurand) [inline src/amo/components/Search/index.js:259]: We should document that because it might not be obvious in the future. I guess that's because the API rejects the invalid category?

**Review by willdurand — APPROVED** [review]
Looks good to me, thanks!

I noticed one other thing but maybe it's fine for now: when we go to a category page, change a filter on the side, change back to how it was initially, the query string remains in the URL.

r+wc

File: src/amo/components/Search/index.js (hunk 17 above)
Comment (bobsilverberg) [reply src/amo/components/Search/index.js:259]: Yes, it does. It sends us a specific message about "invalid category", but I figured it made more sense to just 404 on any 400 than try to look for that specific message. Where would you recommend we document this? Just in a comment above this?

File: src/amo/pages/CategoryPage/index.js (hunk 16 above)
Comment (willdurand) [reply src/amo/pages/CategoryPage/index.js:146]: I think it's fine (and it will be even better with a comment) but we should probably file an issue because for other search parameters, we display the error on the search page, which is kinda nice too.

**Comment by bobsilverberg** [comment]
> Looks good to me, thanks!
> 
> I noticed one other thing but maybe it's fine for now: when we go to a category page, change a filter on the side, change back to how it was initially, the query string remains in the URL.
> 
> r+wc

I'm not too concerned about that. If we really think it's an issue we can file a follow-up, as it would be nice to land this for today's tag.

File: src/amo/components/Search/index.js (hunk 17 above)
Comment (willdurand) [reply src/amo/components/Search/index.js:262]: the comment is perfect.
